crates/bitwarden/src/crypto/enc_string/asymmetric.rs

```diff
@@ -0,0 +1,272 @@
+/// data using AsymmetricCryptoKeys.
+///
+/// The flexibility of the [AsymmEncString] type allows for different encryption algorithms to be used
+/// which is represented by the different variants of the enum.
+///
+/// ## Note
+///
+/// For backwards compatibility we will rarely if ever be able to remove support for decrypting old
+/// variants, but we should be opinionated in which variants are used for encrypting.
+///
+/// ## Variants
+/// - [Rsa2048_OaepSha256_B64](AsymmEncString::Rsa2048_OaepSha256_B64)
+/// - [Rsa2048_OaepSha1_B64](AsymmEncString::Rsa2048_OaepSha1_B64)
+///
+/// ## Serialization
+///
+/// [AsymmEncString] implements [Display] and [FromStr] to allow for easy serialization and uses a
+/// custom scheme to represent the different variants.
+///
+/// The scheme is one of the following schemes:
+/// - `[type].[data]`
+///
+/// Where:
+/// - `[type]`: is a digit number representing the variant.
+/// - `[data]`: is the encrypted data.
+#[derive(Clone)]
+#[allow(unused, non_camel_case_types)]
+pub enum AsymmEncString {
+    /// 3
+    Rsa2048_OaepSha256_B64 { data: Vec<u8> },
+    /// 4
+    Rsa2048_OaepSha1_B64 { data: Vec<u8> },
+    /// 5
+    #[deprecated]
+    Rsa2048_OaepSha256_HmacSha256_B64 { data: Vec<u8>, mac: Vec<u8> },
+    /// 6
+    #[deprecated]
+    Rsa2048_OaepSha1_HmacSha256_B64 { data: Vec<u8>, mac: Vec<u8> },
+}
+
+/// To avoid printing sensitive information, [AsymmEncString] debug prints to `AsymmEncString`.
+impl std::fmt::Debug for AsymmEncString {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("AsymmEncString").finish()
+    }
+}
+
+/// Deserializes an [AsymmEncString] from a string.
+impl FromStr for AsymmEncString {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let (enc_type, parts) = split_enc_string(s);
+        match (enc_type, parts.len()) {
+            ("3", 1) => {
+                let data = from_b64_vec(parts[0])?;
+                Ok(AsymmEncString::Rsa2048_OaepSha256_B64 { data })
+            }
+            ("4", 1) => {
+                let data = from_b64_vec(parts[0])?;
+                Ok(AsymmEncString::Rsa2048_OaepSha1_B64 { data })
+            }
+            #[allow(deprecated)]
+            ("5", 2) => {
+                let data = from_b64_vec(parts[0])?;
+                let mac: Vec<u8> = from_b64_vec(parts[1])?;
+                Ok(AsymmEncString::Rsa2048_OaepSha256_HmacSha256_B64 { data, mac })
+            }
+            #[allow(deprecated)]
+            ("6", 2) => {
+                let data = from_b64_vec(parts[0])?;
+                let mac: Vec<u8> = from_b64_vec(parts[1])?;
+                Ok(AsymmEncString::Rsa2048_OaepSha1_HmacSha256_B64 { data, mac })
+            }
+
+            (enc_type, parts) => Err(EncStringParseError::InvalidType {
+                enc_type: enc_type.to_string(),
+                parts,
+            }
```

Comment:
Do you think it could be confusing to return the same error type for both symmetric and asymmetic? I'm primarily thinking about the cases where we send a symmetric key to asymmetric since it's technically a valid encstring.

Reply:
Yeah fair point, I've split the error into symm/asymm variants.